src/analysis/souffle/taint.dl

```diff
@@ -110,26 +110,68 @@ mayHaveTag(tgt, tag) :-
 // These rules add objects to the set of things that exist (i.e. relations like
 // `isObject(..)`) whenever they are mentioned in other rules.
 
-// Symbols used in edges are access paths
-isAccessPath(x) :- edge(x, _).
-isAccessPath(y) :- edge(_, y).
+// Access paths directly mentioned by other facts.
+.decl isDirectlyMentionedAccessPath(accessPath: AccessPath)
 
-isAccessPath(path) :- claimHasTag(path, _).
-isAccessPath(path) :- mayHaveTag(path, _).
+// Symbols used in edges are access paths
+isDirectlyMentionedAccessPath(x) :- edge(x, _).
+isDirectlyMentionedAccessPath(y) :- edge(_, y).
+
+isDirectlyMentionedAccessPath(path) :- claimHasTag(path, _).
+isDirectlyMentionedAccessPath(path) :- mayHaveTag(path, _).
+
+// Enumerates all potentially-valid indices for any directly-mentioned access path. Because all
+// derived access paths are substrings of directly-mentioned access paths, this gives all
+// potentially-valid indices for all access paths.
+.decl isAccessPathIndex(index: number)
+isAccessPathIndex(0).
+isAccessPathIndex(index + 1) :-
+  isAccessPathIndex(index),
+  isDirectlyMentionedAccessPath(accessPath),
+  index + 1 < strlen(accessPath).
+
+// Access paths split into subpaths at the dot character; this finds the locations at which that
+// split occurs for each access path.
+.decl isMemberBoundary(accessPath: AccessPath, index: number)
+isMemberBoundary(accessPath, index) :-
+  isDirectlyMentionedAccessPath(accessPath),
+  isAccessPathIndex(index),
+  index < strlen(accessPath),
+  substr(accessPath, index, 1) = ".".
+
+// Any access path that was directly mentioned is an access path.
+isAccessPath(accessPath) :- isDirectlyMentionedAccessPath(accessPath).
+
+// Any prefix of a directly mentioned access path that ends in a dot is an access path (not
+// including the dot).
+isAccessPath(substr(accessPath, 0, boundaryIdx)) :-
+  isDirectlyMentionedAccessPath(accessPath),
+  isMemberBoundary(accessPath, boundaryIdx).
 
 // Symbols used in hasTag or mayHaveTag are tags
 isTag(tag) :- claimHasTag(_, tag).
 isTag(tag) :- mayHaveTag(_, tag).
 isTag(tag) :- ownsTag(_, tag).
 isTag(tag) :- saysDowngrades(_, _, tag).
 
+// A pair of (base, member) is in this set if the base access path is a non-trivial subpath of the
+// member access path.
+.decl isMemberOf(base: AccessPath, member: AccessPath)
+
+// The member is a member of the base if it starts with the base plus '.'.
+isMemberOf(base, member) :-
+  isAccessPath(base),
+  isAccessPath(member),
+  strlen(base) + 1 < strlen(member),
+  cat(base, ".") = substr(member, 0, strlen(base) + 1).
+
 // An access path may have a tag if some subpath to a member field has that tag. This allows
 // checking for some inner node in the access path whether any leaf node of that node might have
 // a particular tag.
 mayHaveTag(base, tag) :-
```

Comment:
@bgogul @markww 
I think I missed where this rule got added, but I think this rule is what the rest of this PR is about. We should not have this rule, because it is overly conservative. It essentially computes the join over the labels of the struct and applies this to the entire struct. A particle might only read a part of the struct rather than the entire struct. If it reads just a part of the struct, this will reject more programs than it should.

If we just track labels for completely unfolded paths rather than trying to compute labels for larger structures, the analysis should be more precise and secure. The analysis would also be simpler, it seems.

Reply:
> A particle might only read a part of the struct rather than the entire struct. If it reads just a part of the struct, this will reject more programs than it should.

This is valid point. However, won't we be able to control the precision by defining what `AccessPath` are valid for a particular handle. Suppose we have the following spec:
```
schema Foo 
  a: Text
  b: Text
  c: Text
}
particle FooWriter
  output: writes Foo

particle PartialFooReader
  input: reads Foo {a, b}

recipe R
  foo: create
  FooWriter:
    output: foo
  PartialFooReader
    input: foo
```

We will have the following relations:
```
  edge("output.a", "input.a")
  edge("output.b", "input.b")
 ```
  In particular, there will be no occurrence of `input.c`, and therefore, `isAccessPath("input.c")` will be `false`. Further, `mayHaveTag("input")` will only combine the labels for `input.a` and `input.b`.

WDYT?

Reply:
Shouldn't the two particles expand to the following after we unfold them ?
```
particle FooWriter
     writes: {Foo.a, Foo.b, Foo.c }
  
particle PartialFooReader
     reads: {Foo.a, Foo.b}
```

And if we do this unfolding and the much simpler analysis (including the edges from `edges_from_graph.dl` along with the `isAccessPath(..)` predicates that come indirectly from `bindFlatParticle` and `isFlatParticle`) doesn't this also work? (The result should be that `Foo.a`, `Foo.b`, and `Foo.c` all have the same tags because FooWriter writes them all)

I thought the idea was to have two layers of IR: 1) is similar to arcs protobufs and includes nested structs, 2) has completely unfolded path names.

Reply:
For the Datalog, the plan is to jump right from the original Arcs manifest protos to Datalog facts written in terms of completely unfolded AccessPath names. We will perform the unfolding in the parser that translates from Arcs manifest protos to Datalog facts.

Early on, I wrote a PR that was more similar to what you described, basically directly mapping the proto structures into Datalog facts with a similar structure. @bgogul convinced me that this was not necessary: such complex structures are awkward to express and abstract from within Datalog facts and could be subject to a performance penalty due to Datalog's eager population of intermediate sets. We agreed that the the unfolding you describe should be done within the protobuf translator and that the Souffle datalog analysis should only push abstract state around on a dataflow graph described in Datalog `edge` and `claim` facts.

This PR assumes that edges will be drawn only between leaf fields and that the inner fields of an `AccessPath` will only be referenced in checks. This means that the taint can't flow up and then down the struct as you describe. While the taint flows up for serving checks upon inner nodes, it does not flow down because it does not need to propagate taint from edges connecting to inner nodes to their members, as edges to inner nodes do not exist.

Let me know if you find this argument convincing, or whether I've missed something.

Reply:
A few clarifications:

> Shouldn't the two particles expand to the following after we unfold them ?

Yes, they should expand, but it would happen at the level of the handle connections `input`, `foo`, and `output`.  That is, the edges would be as follows:

```
  edge("output.a", "foo.a")
  edge("output.b", "foo.b")
  edge("output.c", "foo.c")
  edge("foo.b", "foo.b")
  edge("foo.a", "input.a")
  edge("foo.b", "input.b")
```
I had missed the intermediate `foo` in the previous comment. Is this a source of confusion? 

> markww: This PR assumes that edges will be drawn only between leaf fields and that the inner fields of an AccessPath will only be referenced in checks. 

> aferr: If we just track labels for completely unfolded paths rather than trying to compute labels for larger structures, the analysis should be more precise and secure. The analysis would also be simpler, it seems.

It looks like we are aligned? The additional rule in this PR is about inferring a tag for `foo`, `input`, and `output` if needed for a `check` in the IR.  See https://github.com/PolymerLabs/arcs/blob/e7a79c2c885bd090f45c8fb9f8898ba8367f4907/javatests/arcs/core/analysis/testdata/fail_check_on_subpaths.arcs#L9 for an example. These tags would not be propagated automatically. 

> I thought the idea was to have two layers of IR: 1) is similar to arcs protobufs and includes nested structs, 2) has completely unfolded path names.

Yes, we would start with (1) as the input IR. Whether we have (2) as explicit IR is an implementation detail? Ultimately, we want to generate the `edge` relationships in datalog, right?

> For the Datalog, the plan is to jump right from the original Arcs manifest protos to Datalog facts written in terms of completely unfolded AccessPath names. We will perform the unfolding in the parser that translates from Arcs manifest protos to Datalog facts.

In the long run, we want to have an intermediate step, i.e., proto -> C++ data structures -> datalog facts/relations. For the time being, if it turns out to be straightforward to just iterate over the protos and generate the datalog facts directly, we just do that.

Reply:
> Yes, we would start with (1) as the input IR. Whether we have (2) as explicit IR is an implementation detail? Ultimately, we want to generate the edge relationships in datalog, right?

> In the long run, we want to have an intermediate step, i.e., proto -> C++ data structures -> datalog facts/relations. For the time being, if it turns out to be straightforward to just iterate over the protos and generate the datalog facts directly, we just do that.

OK, yes this is just an implementation detail. It doesn't matter if we have these layers.

> See https://github.com/PolymerLabs/arcs/blob/e7a79c2c885bd090f45c8fb9f8898ba8367f4907/javatests/arcs/core/analysis/testdata/fail_check_on_subpaths.arcs#L9 for an example. These tags would not be propagated automatically.

OK. I see that this is needed for passing checks like that and that this code doesn't propagate inward to the leaves. The meaning of tags for outer structures that can't directly be read doesn't make too much sense to me, but I don't think it is harmful, so we can keep it.

Reply:
Hope I was not too in the way!

Reply:
Of course not! Please feel free to bring things up for discussion.

Reply:
Yes, indeed. Program analysis is hard and subtle, better to have the discussion and decide it wasn't an issue than to miss something.